File: cs/Problems/Day21.cs
namespace AOC2023.Problems;

public class Day21 : IProblem<int>
{
    public int Solve(ReadOnlySpan<char> input) => CountNumberOfPlotsInSteps(input);

    private static int CountNumberOfPlotsInSteps(ReadOnlySpan<char> input)
    {
        Span<Range> lineRanges = stackalloc Range[131];
        int size = input.Split(lineRanges, InputReader.NewLine);

        int midpoint = size / 2;
        var nextVisits = new List<Point>(4 * 8);
        var lastVisits = new List<Point> { new(midpoint, midpoint) };

        for (int step = 0; step < 64; step++)
        {
            foreach (var visit in lastVisits)
            {
                foreach (var direction in Point.NavigableDirections)
                {
                    var next = visit.Move(direction);

                    if (next.Y >= size - 1 || next.Y < 0 || next.X >= size - 1 || next.X < 0)
                        continue;

                    var tile = input[lineRanges[next.Y]][next.X];

                    if (tile != '#' && !nextVisits.Contains(next))
                        nextVisits.Add(next);
                }
            }

            lastVisits.Clear();
            lastVisits.AddRange(nextVisits);
            nextVisits.Clear();
        }

        int numberOfPlots = lastVisits.Count;
        return numberOfPlots;
    }

    readonly record struct Point(int X, int Y)
    {
        public static readonly Direction[] NavigableDirections = [Direction.Up, Direction.Down, Direction.Left, Direction.Right];

        public Point Move(Direction direction) =>
            direction switch
            {
                Direction.Up => new(X, Y - 1),
                Direction.Down => new(X, Y + 1),
                Direction.Left => new(X - 1, Y),
                Direction.Right => new(X + 1, Y),
                _ => throw new NotImplementedException()
            };
    }

    enum Direction
    {
        Up,
        Down,
        Left,
        Right
    }
}

File: ts/problems/day21.ts
import { NEW_LINE } from "../inputReader.js";

const countNumberOfPlotsInSteps = (input: string) => countNumberOfPlots(input.split(NEW_LINE));

const countNumberOfPlots = (lines: string[]) => {
  let nextVisits: Point[] = [];
  let lastVisits: Point[] = [{ x: Math.floor(lines.length / 2), y: Math.floor(lines.length / 2) }];

  for (let step = 0; step < 64; step++) {
    for (const visit of lastVisits) {
      for (const direction of NAVIGABLE_DIRECTIONS) {
        const next = movePoint(visit, direction);
        if (isOffset(next, lines.length)) continue;

        const tile = lines[next.y][next.x];
        if (tile != "#" && !nextVisits.some((p) => p.x === next.x && p.y === next.y)) nextVisits.push(next);
      }
    }
    lastVisits = nextVisits;
    nextVisits = [];
  }

  return lastVisits.length;
};

const movePoint = (point: Point, direction: Direction): Point => {
  if (direction == Direction.Up) return { x: point.x, y: point.y - 1 };
  if (direction == Direction.Right) return { x: point.x + 1, y: point.y };
  if (direction == Direction.Down) return { x: point.x, y: point.y + 1 };
  if (direction == Direction.Left) return { x: point.x - 1, y: point.y };
  return point;
};

const isOffset = (point: Point, size: number) => point.x >= size - 1 || point.x < 0 || point.y >= size - 1 || point.y < 0;

enum Direction {
  Up = 0,
  Right = 1,
  Down = 2,
  Left = 3,
}

const NAVIGABLE_DIRECTIONS = [Direction.Up, Direction.Down, Direction.Left, Direction.Right];

type Point = {
  x: number;
  y: number;
};

export default countNumberOfPlotsInSteps;
